src/ScopeProvider/scope.ts

```diff
@@ -0,0 +1,115 @@
+import { atom, type Atom } from 'jotai';
+import { type AnyAtom, type AnyWritableAtom } from './types';
+
+export type Scope = {
+  getAtom: <T extends AnyAtom>(originalAtom: T, fromScoped?: boolean) => T;
+};
+
+export function createScope(
+  atoms: Iterable<AnyAtom>,
+  parentScope: Scope | undefined,
+): Scope {
+  const explicit = new WeakMap<AnyAtom, AnyAtom>();
+  const implicit = new WeakMap<AnyAtom, AnyAtom>();
+  const inherited = new WeakMap<AnyAtom, AnyAtom>();
+  const unscoped = new WeakMap<AnyAtom, AnyAtom>();
```

Comment:
In order of priority descending:
1. `explicit` - map of <originalAtom, scopedCopy>. atoms passed to ScopeProvider are explicit.
1. `implicit` - map of <originalAtom, scopedCopy>. Dependencies of explicit scoped atoms are implicitly scoped.
1. `inherited` - map of <originalAtom, scopedCopy>. Scoped atoms in parent scopes are copied to this scope so they can access scoped atoms in this scope.
1. `unscoped` - map of <originalAtom, scopedCopy>. Unscoped derived atoms are copied so they can access scoped atoms in this scope.